src/databricks/labs/ucx/progress/history.py

```diff
@@ -0,0 +1,240 @@
+from __future__ import annotations
+import dataclasses
+import datetime as dt
+import json
+import logging
+from collections.abc import Iterable, Sequence
+from typing import ClassVar, Protocol, TypeVar, Generic, Any, get_type_hints
+
+from databricks.labs.lsql.backends import SqlBackend
+
+from databricks.labs.ucx.framework.owners import Ownership
+from databricks.labs.ucx.framework.utils import escape_sql_identifier
+from databricks.labs.ucx.progress.install import Historical
+
+logger = logging.getLogger(__name__)
+
+
+class DataclassWithIdAttributes(Protocol):
+    __dataclass_fields__: ClassVar[dict[str, Any]]
+
+    __id_attributes__: ClassVar[tuple[str, ...]]
+    """The names of attributes (can be dataclass fields or ordinary properties) that make up the object identifier.
+
+    All attributes must be (non-optional) strings.
+    """
+
+
+Record = TypeVar("Record", bound=DataclassWithIdAttributes)
+
+
+class HistoricalEncoder(Generic[Record]):
+    """An encoder for dataclasses that will be stored in our history log.
+
+    Our history records are designed with several goals in mind:
+     - Records can be of different types, meaning we have to store heterogenous types using a homogenous schema.
+     - We partially shred the records to allow for easier SQL-based querying.
+     - Flexibility from the start, because schema changes in the future will be very difficult.
+     - Records will be shared (and queried) across workspaces with different versions of UCX installed.
+
+    With this in mind:
+     - We have an 'object-type' (discriminator) field for holding the type of the record we're storing. This allows for
+       type-specific logic when querying.
+     - We have a (type-specific) field for storing the business key (identifier) of each record.
+     - We have special handling of the 'failures' attribute that is normally present in the types we are storing.
```

Comment:
```suggestion
     - We have special handling of the 'failures' attribute that is sometimes present in the types we are storing.
```

Note on the failures: it is present when we "assess" objects not when we "crawl" objects. At least, that is the common discriminator. In practice, we are not so strict about this

Reply:
I've updated the wording here.